### PR TITLE
Fuchsia

### PR DESCRIPTION
**Pull request by trishankatdatadog**
Merge Fuchsia [fork](https://fuchsia.googlesource.com/third_party/go-tuf) with master.

Cc @erickt @raggi @joshuaseaton

### CODE REVIEW

File: data/types.go [hunk 1]
```diff
@@ -90,42 +90,52 @@ type Role struct {
 
 type Files map[string]FileMeta
 
+type FileMeta struct {
+	Length int64            `json:"length",omitempty`
+	Hashes Hashes           `json:"hashes",ompitempty`
```
Comment (joshuagl) [inline data/types.go:180]:
```suggestion
	Hashes Hashes           `json:"hashes",omitempty`
```
I don't speak Go, but this looks like a typo?

**Comment by joshuagl** [comment]
The CI failure is because client/interop_test.go clones a copy of tuf, but does not ensure there's a version of securesystemslib installed for tuf to use: https://github.com/theupdateframework/go-tuf/blob/bf368c57efac4ff6b5976e3669a391b1ca186f48/client/interop_test.go#L108-L116

**Comment by trishankatdatadog** [comment]
> The CI failure is because client/interop_test.go clones a copy of tuf, but does not ensure there's a version of securesystemslib installed for tuf to use:

Yeah, I saw that, but have not reproduced on my machine yet, not sure why, will look deeper tonight

**Comment by trishankatdatadog** [comment]
> Yeah, I saw that, but have not reproduced on my machine yet, not sure why, will look deeper tonight

Ok, it's because locally I have a copy of `securesystemslib`. I think it's best to let Travis handle this than solve this at the Go tests level. Development setup is a different story. I'll fix it tomorrow.

**Comment by trishankatdatadog** [comment]
I can't reproduce locally what is going wrong on Travis, and I'm running out of time to continue looking at this. Could someone please help take a look?

**Comment by shibumi** [comment]
@trishankatdatadog The reason why it's working for you locally is, you are using another version of the TUF python implementation. If I use the TUF develop branch, the test works for me:

```
❯ pip install https://github.com/theupdateframework/tuf/archive/develop.tar.gz --user --upgrade
Collecting https://github.com/theupdateframework/tuf/archive/develop.tar.gz
  Downloading https://github.com/theupdateframework/tuf/archive/develop.tar.gz
     - 4.6 MB 3.0 MB/s
Requirement already satisfied, skipping upgrade: iso8601>=0.1.12 in ./.local/lib/python3.8/site-packages (from tuf==0.12.2) (0.1.12)
Requirement already satisfied, skipping upgrade: requests>=2.19.1 in /usr/lib/python3.8/site-packages (from tuf==0.12.2) (2.24.0)
Requirement already satisfied, skipping upgrade: six>=1.11.0 in /usr/lib/python3.8/site-packages (from tuf==0.12.2) (1.15.0)
Requirement already satisfied, skipping upgrade: securesystemslib>=0.15.0 in ./.local/lib/python3.8/site-packages (from tuf==0.12.2) (0.15.0)
Requirement already satisfied, skipping upgrade: chardet>=3.0.2 in /usr/lib/python3.8/site-packages (from requests>=2.19.1->tuf==0.12.2) (3.0.4)
Requirement already satisfied, skipping upgrade: idna>=2.5 in /usr/lib/python3.8/site-packages (from requests>=2.19.1->tuf==0.12.2) (2.10)
Requirement already satisfied, skipping upgrade: urllib3>=1.21.1 in /usr/lib/python3.8/site-packages (from requests>=2.19.1->tuf==0.12.2) (1.25.10)
Requirement already satisfied, skipping upgrade: python-dateutil>=2.8.0 in /usr/lib/python3.8/site-packages (from securesystemslib>=0.15.0->tuf==0.12.2) (2.8.1)
Using legacy setup.py install for tuf, since package 'wheel' is not installed.
Installing collected packages: tuf
  Attempting uninstall: tuf
    Found existing installation: tuf 0.12.2
    Uninstalling tuf-0.12.2:
      Successfully uninstalled tuf-0.12.2
    Running setup.py install for tuf ... done
Successfully installed tuf-0.12.2
```

```
chris motoko ~/go/src/github.com/theupdateframework/go-tuf 1:27:01 fa60d5e5 remotes/origin/fuchsia  
❯ go test ./...
ok      github.com/theupdateframework/go-tuf    (cached)
ok      github.com/theupdateframework/go-tuf/client     1.478s
ok      github.com/theupdateframework/go-tuf/client/leveldbstore        (cached)
ok      github.com/theupdateframework/go-tuf/client/python_interop      0.609s
?       github.com/theupdateframework/go-tuf/cmd/tuf    [no test files]
?       github.com/theupdateframework/go-tuf/cmd/tuf-client     [no test files]
ok      github.com/theupdateframework/go-tuf/data       (cached)
ok      github.com/theupdateframework/go-tuf/encrypted  (cached)
ok      github.com/theupdateframework/go-tuf/sign       (cached)
ok      github.com/theupdateframework/go-tuf/util       (cached)
ok      github.com/theupdateframework/go-tuf/verify     (cached)
```

The corresponding test will still fire an exception:
```
ERROR:root:HTTP error for root version 2
Traceback (most recent call last):
  File "/home/chris/.local/lib/python3.8/site-packages/tuf/client/updater.py", line 1145, in _update_root_metadata
    self._update_metadata('root', DEFAULT_ROOT_UPPERLENGTH,
  File "/home/chris/.local/lib/python3.8/site-packages/tuf/client/updater.py", line 1781, in _update_metadata
    self._get_metadata_file(metadata_role, remote_filename,
  File "/home/chris/.local/lib/python3.8/site-packages/tuf/client/updater.py", line 1598, in _get_metadata_file
    raise tuf.exceptions.NoWorkingMirrorError(file_mirror_errors)
tuf.exceptions.NoWorkingMirrorError: No working mirror was found:
  '127.0.0.1:37081': HTTPError('404 Client Error: Not Found for url: http://127.0.0.1:37081/consistent-snapshot-false/repository/2.root.json')
ERROR:root:HTTP error for root version 2
Traceback (most recent call last):
  File "/home/chris/.local/lib/python3.8/site-packages/tuf/client/updater.py", line 1145, in _update_root_metadata
    self._update_metadata('root', DEFAULT_ROOT_UPPERLENGTH,
  File "/home/chris/.local/lib/python3.8/site-packages/tuf/client/updater.py", line 1781, in _update_metadata
    self._get_metadata_file(metadata_role, remote_filename,
  File "/home/chris/.local/lib/python3.8/site-packages/tuf/client/updater.py", line 1598, in _get_metadata_file
    raise tuf.exceptions.NoWorkingMirrorError(file_mirror_errors)
tuf.exceptions.NoWorkingMirrorError: No working mirror was found:
  '127.0.0.1:37081': HTTPError('404 Client Error: Not Found for url: http://127.0.0.1:37081/consistent-snapshot-true/repository/2.root.json')
OK: 1 passed
--- PASS: Test (0.51s)
Process finished with exit code 0
```

but the test will pass at least.

I suggest you create a requirements.txt file or you specify the whole path in the travis-ci file for the TUF package.

**Comment by shibumi** [comment]
This patch should work:
```diff
From 37c9bf5a0d7320e3ff350d652b8fdd235369d079 Mon Sep 17 00:00:00 2001
From: Christian Rebischke <chris@shibumi.dev>
Date: Thu, 30 Jul 2020 01:44:31 +0200
Subject: [PATCH] Use develop branch for TUF

We need to pull in the latest changes of TUF.
Otherwise our travis builds will fail.

Signed-off-by: Christian Rebischke <chris@shibumi.dev>
---
 .travis.yml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/.travis.yml b/.travis.yml
index 76be434..6ffc0c8 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -7,7 +7,7 @@ install: true
 env:
   - GO111MODULE=on
 script:
-  - pip install iso8601 requests securesystemslib six tuf
+  - pip install iso8601 requests securesystemslib six https://github.com/theupdateframework/tuf/archive/develop.tar.gz
   - go test -race ./...

 notifications:
--
2.28.0
```

**Comment by trishankatdatadog** [comment]
Thanks, Christian! It seems to be a bug where no official release of `tuf` has yet caught up with recent changes in `securesystemslib` (looks like the `storage_backend` stuff) @joshuagl @lukpueh

**Comment by joshuagl** [comment]
> Thanks, Christian! It seems to be a bug where no official release of `tuf` has yet caught up with recent changes in `securesystemslib` (looks like the `storage_backend` stuff) @joshuagl @lukpueh

Working on it https://github.com/theupdateframework/tuf/pull/1094

**Comment by erickt** [comment]
@trishankatdatadog LGTM. There's probably some cleanup we can do, but we could land this PR first, and clean up in future PRs.

**Comment by shibumi** [comment]
@erickt can you open issues for the cleanups you are seeing? Would be super helpful, then maybe I could clean this up in later PRs.

**Pull request merged**